File: flutter/lib/domain/ship_navigation/find_path.dart
import 'package:ship_conquest/domain/ship_navigation/distances.dart';
import 'package:ship_conquest/domain/ship_navigation/node.dart';
import 'package:ship_conquest/domain/ship_navigation/utils.dart';
import 'package:ship_conquest/domain/space/map_position.dart';


List<MapPosition> findShortestPath(List<MapPosition> map, MapPosition start,
    MapPosition end, MapPosition influencePoint, int safetyRadius, int mapSize) {
  final double startH = calculateHeuristic(start, end, influencePoint);
  final Node startNode = Node(position: start, h: startH, f: startH);
  final List<Node> foundNodes = List.empty(growable: true);
  foundNodes.add(startNode);
  final List<Node> exploredNodes = List.empty(growable: true);

  while (foundNodes.isNotEmpty) {
    final Node currNode = foundNodes.reduce((firstNode, secondNode) =>
        firstNode.f < secondNode.f ? firstNode : secondNode);
    if (currNode.position == end) {
      return reconstructPath(currNode);
    }
    foundNodes.remove(currNode);
    exploredNodes.add(currNode);
    final List<Node> neighbors = calculateNeighbours(currNode, map, safetyRadius, mapSize, exploredNodes);

    for(Node currNeighbour in neighbors) {
      if(exploredNodes.contains(currNeighbour)) {
        continue;
      }

      final double tempG = currNode.g + calculateEuclideanDistance(currNode.position, currNeighbour.position);

      final Node neighbourNode = foundNodes.firstWhere((node) => node.position == currNeighbour.position, orElse: () => currNeighbour);

      if(tempG < neighbourNode.g || !foundNodes.contains(currNeighbour)) {
        final double h = calculateHeuristic(currNeighbour.position, end, influencePoint);
        final Node copyNode = currNeighbour.copyWith(g: tempG, h: h, f: tempG + h, parent: currNode);

        Node? existingNode;
        try {
          existingNode = foundNodes.firstWhere((node) => node.position == copyNode.position);
        } catch (error) {
          existingNode = null;
        }
        if(existingNode != null) {
          if(copyNode.f < existingNode.f) {
            foundNodes.add(copyNode);
            foundNodes.remove(existingNode);
          }
        } else {
          foundNodes.add(copyNode);
        }
      }
    }
  }
  return List.empty(growable: false);
}

File: jvm/ship-conquest/src/main/kotlin/com/example/shipconquest/domain/ship_navigation/findPath.kt
package com.example.shipconquest.domain.ship_navigation

import com.example.shipconquest.domain.Position

/**
 *  Using the A* algorithm, find the shortest path avoiding the existing isles and using the influence point to modify the result
 *  @param start the starting point
 *  @param end the end point
 *  @param influencePoint the point to influence the path
 *  @return list of positions with the full path
 */
fun findShortestPath(
    map: List<Position>,
    start: Position,
    end: Position,
    influencePoint: Position,
    safetyRadius: Int,
    mapSize: Int
): List<Position> {
    val startH = calculateHeuristic(start, end, influencePoint)
    val startNode = Node(position = start, h = startH, f = startH)
    val foundNodes = mutableSetOf(startNode)
    val exploredNodes = mutableSetOf<Node>()

    while (foundNodes.isNotEmpty()) {
        val currNode = foundNodes.minByOrNull { node -> node.f }
            ?: return emptyList() //find the one with the least f
        if (currNode.position == end) {
            return reconstructPath(currNode)
        }
        foundNodes.remove(currNode) //remove the node from the list
        exploredNodes.add(currNode)
        val neighbors = calculateNeighbours(currNode, map, safetyRadius, mapSize, exploredNodes)

        for (currNeighbour in neighbors) {
            if (currNeighbour in exploredNodes) {
                continue
            }

            val tempG = currNode.g + calculateEuclideanDistance(currNode.position, currNeighbour.position)

            val neighbourNode = foundNodes.find { it.position == currNeighbour.position }
                ?: currNeighbour

            if (tempG < neighbourNode.g || currNeighbour !in foundNodes) {
                // calculate f by the sum of g and h
                val h = calculateHeuristic(currNeighbour.position, end, influencePoint)
                val copyNode = currNeighbour.copy(g = tempG, h = h, f = tempG + h, parent = currNode)

                //optimization to avoid node repetition with worse f value, reducing the size of the list
                val existingNode = foundNodes.find { node -> node.position == copyNode.position }
                if (existingNode != null) {
                    if (copyNode.f < existingNode.f) {
                        foundNodes.add(copyNode)
                        foundNodes.remove(existingNode)
                    }
                } else {
                    foundNodes.add(copyNode)
                }
            }
        }
    }
    return emptyList()
}